### PR TITLE
ISPN-5542 Shorten interceptor stack for local caches

### PR DESCRIPTION
**Pull request by rvansa**
https://issues.jboss.org/browse/ISPN-5542

### CODE REVIEW

**Comment by Sanne** [comment]
Hi Radim, did you consider a 2lc implementation which simply doesn't use the interceptor chain at all?

In case we want to go for the "slimming" approach, I'm with you on removing some like transactions etc, but I doubt it should be permitted to completely remove the statistics one.. isn't that going a bit too far? Even when tuning a local-only cache, being able to enable statistics seems very useful, while the benefit of removing that one is probably very hard to find.

+1 to remove it if you're thinking to remove them all eventually, and I just happen to look at the first step.

Rather than log a warning in the case someone enables statistics when it's not enabled, should it not throw an exception?

**Comment by rvansa** [comment]
Yes, I was posting about local cache without interceptor chain to infinispan-dev.

It's always just option, and by default the statistics are there. Once you have your settings tuned and don't plan to look on them anymore, you can remove them. If I recall correctly I've seen benefit of about 5-10% increased throughput of get operations.

I can't remove all interceptors without replacing some of the functionality; that's why the interceptors are there.

You're right, incompatible configuration should end up with CacheConfigurationException.

**Comment by Sanne** [comment]
ok great. If you can change the configuration validation to throw an exception I'll test & merge it.

**Comment by rvansa** [comment]
@Sanne updated, thanks

File: core/src/main/java/org/infinispan/configuration/cache/JMXStatisticsConfigurationBuilder.java [hunk 1]
```diff
@@ -44,8 +50,20 @@ public JMXStatisticsConfigurationBuilder enabled(boolean enabled) {
       return this;
    }
 
+   public JMXStatisticsConfigurationBuilder available(boolean available) {
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/configuration/cache/JMXStatisticsConfigurationBuilder.java]: No javadoc?

**Comment by Sanne** [comment]
:+1: 
Merged! Looking forward for more on this theme

File: core/src/main/java/org/infinispan/configuration/cache/TransactionConfigurationBuilder.java [hunk 2]
```diff
@@ -259,6 +259,11 @@ public TransactionConfigurationBuilder transactionProtocol(TransactionProtocol t
       return this;
    }
 
+   public TransactionConfigurationBuilder notifications(boolean enabled) {
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/configuration/cache/TransactionConfigurationBuilder.java]: Builder should have javadoc as well.

File: core/src/main/java/org/infinispan/util/logging/Log.java [hunk 3]
```diff
@@ -1,5 +1,26 @@
...
+import java.io.IOException;
+import java.lang.reflect.Method;
+import java.net.URL;
+import java.nio.channels.FileChannel;
+import java.security.Permission;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.CountDownLatch;
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/util/logging/Log.java]: Imports being shuffled around...

File: core/src/main/java/org/infinispan/util/logging/Log.java [hunk 4]
```diff
@@ -1348,4 +1352,7 @@ void asyncStoreShutdownTimeoutTooHigh(long configuredAsyncStopTimeout,
 
    @Message(value = "Cannot remove cache configuration '%s' because it is in use", id = 371)
    IllegalStateException configurationInUse(String configurationName);
+
+   @Message(value = "Statistics are enabled while these are not available.", id = 372)
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/util/logging/Log.java]: It's not very clear from the error message that 'available' and 'enabled' are two different, and related, attributes.

File: core/src/main/resources/schema/infinispan-config-8.0.xsd [hunk 5]
```diff
@@ -763,6 +763,13 @@
         </xs:documentation>
       </xs:annotation>
     </xs:attribute>
+    <xs:attribute name="notifications" type="xs:boolean" default="true">
+      <xs:annotation>
+        <xs:documentation>
+          Enables or disables triggering transactional notifications on cache listeners. By default is enabled.
+        </xs:documentation>
+      </xs:annotation>
+    </xs:attribute>
```
Comment (danberindei) [inline core/src/main/resources/schema/infinispan-config-8.0.xsd]: No XSD change necessary for statistics-available?

File: core/src/main/java/org/infinispan/interceptors/CacheMgmtInterceptor.java [hunk 6]
```diff
@@ -8,9 +8,9 @@
 
 import org.infinispan.commands.FlagAffectedCommand;
 import org.infinispan.commands.read.AbstractDataCommand;
+import org.infinispan.commands.read.GetAllCommand;
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/interceptors/CacheMgmtInterceptor.java]: Unnecessary import shuffling...

**Comment by Sanne** [comment]
@danberindei sorry, we had a race condition and I merged this before you started looking.

I did revert all import shuffling before merging though ;-)
Could someone send an additional PR to improve the javadocs, the error message, etc.. ?

**Comment by danberindei** [comment]
I would have also liked to see the attributes added to the "complete" XML configuration in `named-cache-test.xml`

**Comment by danberindei** [comment]
@Sanne I was hoping you'd add a comment yourself about the imports :)

**Comment by danberindei** [comment]
@Sanne Maybe it's good you merged it now, I would have asked @rvansa to provide some concrete benchmark results as well :)

**Comment by rvansa** [comment]
I'll add the javadoc and xsd docs right away...

@danberindei https://gist.github.com/rvansa/ede7933944f4acd54066

**Comment by rvansa** [comment]
@Sanne @danberindei Follow-up: https://github.com/infinispan/infinispan/pull/3560

**Comment by tristantarrant** [comment]
@rvansa , as @danberindei said, any chance for some benchmark results ? Any changes in memory allocation (I can't remember if the stats code does any significant allocations)

**Comment by rvansa** [comment]
@tristantarrant A micro-benchmark results are here: https://gist.github.com/rvansa/ede7933944f4acd54066
There shouldn't be any changes in memory allocation, just invoking less megamorphic calls and stack traversal.

**Pull request closed without merging**